zstd/seqdec_amd64.s

```diff
@@ -0,0 +1,600 @@
+	MOVQ    s+0(FP), DI
+	MOVQ    seqVals_mo(seqs), BX            // mo
+	MOVQ    seqVals_ll(seqs), CX            // ll
+	MOVQ    R14, DX                         // moB (from the ofState before its update)
+	SHRQ    $8, DX
+	MOVBQZX DL, DX
+	LEAQ    sequenceDecs_prevOffset(DI), DI // DI = &s.prevOffset[0]
+
+	// if offsetB > 1 {
+	//     s.prevOffset[2] = s.prevOffset[1]
+	//     s.prevOffset[1] = s.prevOffset[0]
+	//     s.prevOffset[0] = offset
+	//     return offset
+	// }
+	CMPQ DX, $1
+	JBE  offsetB_1_or_0
+
+	MOVQ 8(DI), AX
+	MOVQ AX, 16(DI) // s.prevOffset[2] = s.prevOffset[1]
+	MOVQ 0(DI), AX
+	MOVQ AX, 8(DI)  // s.prevOffset[1] = s.prevOffset[0]
+	MOVQ BX, 0(DI)  // s.prevOffset[0] = offset
+
+	MOVQ BX, BX
+	JMP  adjust_offsets_end
+
+offsetB_1_or_0:
+	// if litLen == 0 {
+	//     offset++
+	// }
+	LEAQ    1(BX), AX // AX = offset + 1
+	TESTQ   CX, CX
+	CMOVQEQ AX, BX    // offset++ if litLen == 0
+
+	// if offset == 0 {
+	//     return s.prevOffset[0]
+	// }
+	TESTQ BX, BX
+	JNZ   offset_nonzero
+	MOVQ  0(DI), BX
+	JMP   adjust_offsets_end
+
+offset_nonzero:
+	// Note: at this point CX (litLen) and DX (offsetB) are free
+
+	// var temp int
+	// if offset == 3 {
+	//     temp = s.prevOffset[0] - 1
+	// } else {
+	//     temp = s.prevOffset[offset]
+	// }
+	//
+	// this if got transformed into:
+	//
+	// ofs   := offset
+	// shift := 0
+	// if offset == 3 {
+	//     ofs   = 0
+	//     shift = -1
+	// }
+	// temp := s.prevOffset[ofs] + shift
+
+	MOVQ    BX, CX   // ofs_other   = offset
+	XORQ    DX, DX   // shift_other = 0
+	XORQ    R14, R14 // ofs_3       = 0
+	MOVQ    $-1, R15 // shift_3     = -1
+	CMPQ    BX, $3
+	CMOVQEQ R14, CX
+	CMOVQEQ R15, DX
+
+	ADDQ 0(DI)(CX*8), DX // DX is temp
+
+	// if temp == 0 {
+	//     temp = 1
+	// }
+	MOVQ    $1, CX
+	CMOVQEQ CX, DX
+
+	// if offset != 1 {
+	//     s.prevOffset[2] = s.prevOffset[1]
+	// }
+	CMPQ BX, $1
+	JZ   skip
+	MOVQ 8(DI), AX
+	MOVQ AX, 16(DI) // s.prevOffset[2] = s.prevOffset[1]
+
+skip:
+	// s.prevOffset[1] = s.prevOffset[0]
+	// s.prevOffset[0] = temp
+	MOVQ 0(DI), AX
+	MOVQ AX, 8(DI) // s.prevOffset[1] = s.prevOffset[0]
+	MOVQ DX, 0(DI) // s.prevOffset[0] = temp
+
+	// return temp
+	MOVQ DX, BX
+
+adjust_offsets_end:
+
+check_triple:
+
+	MOVQ BX, seqVals_mo(seqs) // BX - mo
+	MOVQ seqVals_ml(seqs), CX // CX - ml
+	MOVQ seqVals_ll(seqs), DX // DX - ll
+
+	MOVQ s+0(FP), DI
+	LEAQ 0(CX)(DX*1), AX
+	ADDQ AX, sequenceDecs_seqSize(DI) // s.seqSize += ml + ll
+
+	MOVQ ctx+16(FP), DI
+	SUBQ DX, decodeAsmContext_litRemain(DI) // ctx.litRemain -= ll
+
+/*
+		if ml > maxMatchLen {
+			return fmt.Errorf("match len (%d) bigger than max allowed length", ml)
+		}
+    */
+	CMPQ CX, $maxMatchLen
+	JA   error_match_len_too_big
+/*
+		if mo == 0 && ml > 0 {
+			return fmt.Errorf("zero matchoff and matchlen (%d) > 0", ml)
+		}
+    */
+	XORQ  AX, AX
+	TESTQ BX, BX
+	SETEQ DL
+	CMPQ  CX, $0
+	SETHI AL
+	ANDQ  DX, AX
+	TESTQ AX, AX
+	JNZ   error_match_len_ofs_mismatch
```

Comment:
Could you try out this. Should be much faster, since the first branch will be taken all the time on valid input.
```suggestion
	TESTQ BX, BX // if mo == 0
	JNZ   match_len_ofs_ok // Taken 100% of the time on valid input
	TESTQ CX, CX // if ml != 0
	JNZ   error_match_len_ofs_mismatch

match_len_ofs_ok:
```

Reply:
Right, `ml` is always non-negative (despite that type `int` is used).

Reply:
I compared the current performance -- after removing casts, `and`s and use conditional jumps -- with the initial perf (with BMI instructions). It's better in most cases. Thank you for such a great review. :)

```
benchmark                                                                 old ns/op     new ns/op     delta
BenchmarkDecoder_DecoderSmall/kppkn.gtb.zst-16                            6080641       5721672       -5.90%
BenchmarkDecoder_DecoderSmall/geo.protodata.zst-16                        1103300       1019169       -7.63%
BenchmarkDecoder_DecoderSmall/plrabn12.txt.zst-16                         19351492      18560199      -4.09%
BenchmarkDecoder_DecoderSmall/lcet10.txt.zst-16                           14446995      13580849      -6.00%
BenchmarkDecoder_DecoderSmall/asyoulik.txt.zst-16                         4569453       4299068       -5.92%
BenchmarkDecoder_DecoderSmall/alice29.txt.zst-16                          6340637       5978242       -5.72%
BenchmarkDecoder_DecoderSmall/html_x_4.zst-16                             1664982       1572635       -5.55%
BenchmarkDecoder_DecoderSmall/paper-100k.pdf.zst-16                       248301        242900        -2.18%
BenchmarkDecoder_DecoderSmall/fireworks.jpeg.zst-16                       129965        128953        -0.78%
BenchmarkDecoder_DecoderSmall/urls.10K.zst-16                             15227573      14628889      -3.93%
BenchmarkDecoder_DecoderSmall/html.zst-16                                 1192996       1123798       -5.80%
BenchmarkDecoder_DecoderSmall/comp-data.bin.zst-16                        87100         84376         -3.13%
BenchmarkDecoder_DecodeAll/kppkn.gtb.zst-16                               757857        726832        -4.09%
BenchmarkDecoder_DecodeAll/geo.protodata.zst-16                           129944        124470        -4.21%
BenchmarkDecoder_DecodeAll/plrabn12.txt.zst-16                            2381303       2242399       -5.83%
BenchmarkDecoder_DecodeAll/lcet10.txt.zst-16                              1767007       1669678       -5.51%
BenchmarkDecoder_DecodeAll/asyoulik.txt.zst-16                            583827        557384        -4.53%
BenchmarkDecoder_DecodeAll/alice29.txt.zst-16                             789483        750024        -5.00%
BenchmarkDecoder_DecodeAll/html_x_4.zst-16                                273257        259808        -4.92%
BenchmarkDecoder_DecodeAll/paper-100k.pdf.zst-16                          27347         25893         -5.32%
BenchmarkDecoder_DecodeAll/fireworks.jpeg.zst-16                          11304         11295         -0.08%
BenchmarkDecoder_DecodeAll/urls.10K.zst-16                                1808791       1724402       -4.67%
BenchmarkDecoder_DecodeAll/html.zst-16                                    152580        145352        -4.74%
BenchmarkDecoder_DecodeAll/comp-data.bin.zst-16                           11377         10684         -6.09%
BenchmarkDecoder_DecodeAllFiles/Mark.Twain-Tom.Sawyer.txt/fastest-16      1731442       1639777       -5.29%
BenchmarkDecoder_DecodeAllFiles/Mark.Twain-Tom.Sawyer.txt/default-16      1868472       1762828       -5.65%
BenchmarkDecoder_DecodeAllFiles/Mark.Twain-Tom.Sawyer.txt/better-16       1724697       1625604       -5.75%
BenchmarkDecoder_DecodeAllFiles/Mark.Twain-Tom.Sawyer.txt/best-16         1793428       1758123       -1.97%
BenchmarkDecoder_DecodeAllFiles/e.txt/fastest-16                          9197          9185          -0.13%
BenchmarkDecoder_DecodeAllFiles/e.txt/default-16                          409277        384598        -6.03%
BenchmarkDecoder_DecodeAllFiles/e.txt/better-16                           278491        267225        -4.05%
BenchmarkDecoder_DecodeAllFiles/e.txt/best-16                             158901        158818        -0.05%
BenchmarkDecoder_DecodeAllFiles/fse-artifact3.bin/fastest-16              3525          3413          -3.18%
BenchmarkDecoder_DecodeAllFiles/fse-artifact3.bin/default-16              3377          3253          -3.67%
BenchmarkDecoder_DecodeAllFiles/fse-artifact3.bin/better-16               3790          3736          -1.42%
BenchmarkDecoder_DecodeAllFiles/fse-artifact3.bin/best-16                 9414          9398          -0.17%
BenchmarkDecoder_DecodeAllFiles/gettysburg.txt/fastest-16                 5219          5285          +1.26%
BenchmarkDecoder_DecodeAllFiles/gettysburg.txt/default-16                 8473          8314          -1.88%
BenchmarkDecoder_DecodeAllFiles/gettysburg.txt/better-16                  8423          8346          -0.91%
BenchmarkDecoder_DecodeAllFiles/gettysburg.txt/best-16                    9274          9092          -1.96%
BenchmarkDecoder_DecodeAllFiles/html.txt/fastest-16                       104606        98763         -5.59%
BenchmarkDecoder_DecodeAllFiles/html.txt/default-16                       112972        105033        -7.03%
BenchmarkDecoder_DecodeAllFiles/html.txt/better-16                        99400         95713         -3.71%
BenchmarkDecoder_DecodeAllFiles/html.txt/best-16                          108115        112985        +4.50%
BenchmarkDecoder_DecodeAllFiles/pi.txt/fastest-16                         9185          9188          +0.03%
BenchmarkDecoder_DecodeAllFiles/pi.txt/default-16                         401783        390653        -2.77%
BenchmarkDecoder_DecodeAllFiles/pi.txt/better-16                          270118        262294        -2.90%
BenchmarkDecoder_DecodeAllFiles/pi.txt/best-16                            157041        158702        +1.06%
BenchmarkDecoder_DecodeAllFiles/pngdata.bin/fastest-16                    29755         30341         +1.97%
BenchmarkDecoder_DecodeAllFiles/pngdata.bin/default-16                    32236         32116         -0.37%
BenchmarkDecoder_DecodeAllFiles/pngdata.bin/better-16                     26462         26327         -0.51%
BenchmarkDecoder_DecodeAllFiles/pngdata.bin/best-16                       30678         30669         -0.03%
BenchmarkDecoder_DecodeAllFiles/sharnd.out/fastest-16                     9185          9180          -0.05%
BenchmarkDecoder_DecodeAllFiles/sharnd.out/default-16                     9179          9173          -0.07%
BenchmarkDecoder_DecodeAllFiles/sharnd.out/better-16                      9182          9180          -0.02%
BenchmarkDecoder_DecodeAllFiles/sharnd.out/best-16                        9172          9178          +0.07%
BenchmarkDecoder_DecodeAllFilesP/Mark.Twain-Tom.Sawyer.txt/fastest-16     234753        232549        -0.94%
BenchmarkDecoder_DecodeAllFilesP/Mark.Twain-Tom.Sawyer.txt/default-16     267925        263255        -1.74%
BenchmarkDecoder_DecodeAllFilesP/Mark.Twain-Tom.Sawyer.txt/better-16      238474        236399        -0.87%
BenchmarkDecoder_DecodeAllFilesP/Mark.Twain-Tom.Sawyer.txt/best-16        251072        250560        -0.20%
BenchmarkDecoder_DecodeAllFilesP/e.txt/fastest-16                         1039          1044          +0.48%
BenchmarkDecoder_DecodeAllFilesP/e.txt/default-16                         64429         64166         -0.41%
BenchmarkDecoder_DecodeAllFilesP/e.txt/better-16                          41588         41380         -0.50%
BenchmarkDecoder_DecodeAllFilesP/e.txt/best-16                            20262         20260         -0.01%
BenchmarkDecoder_DecodeAllFilesP/fse-artifact3.bin/fastest-16             766           689           -9.94%
BenchmarkDecoder_DecodeAllFilesP/fse-artifact3.bin/default-16             750           720           -3.88%
BenchmarkDecoder_DecodeAllFilesP/fse-artifact3.bin/better-16              687           666           -3.00%
BenchmarkDecoder_DecodeAllFilesP/fse-artifact3.bin/best-16                1034          1029          -0.48%
BenchmarkDecoder_DecodeAllFilesP/gettysburg.txt/fastest-16                1037          1065          +2.70%
BenchmarkDecoder_DecodeAllFilesP/gettysburg.txt/default-16                1579          1596          +1.08%
BenchmarkDecoder_DecodeAllFilesP/gettysburg.txt/better-16                 1558          1563          +0.32%
BenchmarkDecoder_DecodeAllFilesP/gettysburg.txt/best-16                   1876          1897          +1.12%
BenchmarkDecoder_DecodeAllFilesP/html.txt/fastest-16                      22262         21861         -1.80%
BenchmarkDecoder_DecodeAllFilesP/html.txt/default-16                      23739         23286         -1.91%
BenchmarkDecoder_DecodeAllFilesP/html.txt/better-16                       21604         21706         +0.47%
BenchmarkDecoder_DecodeAllFilesP/html.txt/best-16                         24298         22877         -5.85%
BenchmarkDecoder_DecodeAllFilesP/pi.txt/fastest-16                        1035          1028          -0.68%
BenchmarkDecoder_DecodeAllFilesP/pi.txt/default-16                        64554         63773         -1.21%
BenchmarkDecoder_DecodeAllFilesP/pi.txt/better-16                         41005         40604         -0.98%
BenchmarkDecoder_DecodeAllFilesP/pi.txt/best-16                           20175         20164         -0.05%
BenchmarkDecoder_DecodeAllFilesP/pngdata.bin/fastest-16                   5675          5739          +1.13%
BenchmarkDecoder_DecodeAllFilesP/pngdata.bin/default-16                   5618          5653          +0.62%
BenchmarkDecoder_DecodeAllFilesP/pngdata.bin/better-16                    5108          5039          -1.35%
BenchmarkDecoder_DecodeAllFilesP/pngdata.bin/best-16                      4298          4367          +1.61%
BenchmarkDecoder_DecodeAllFilesP/sharnd.out/fastest-16                    1050          1041          -0.86%
BenchmarkDecoder_DecodeAllFilesP/sharnd.out/default-16                    1045          1038          -0.67%
BenchmarkDecoder_DecodeAllFilesP/sharnd.out/better-16                     1035          1038          +0.29%
BenchmarkDecoder_DecodeAllFilesP/sharnd.out/best-16                       1062          1051          -1.04%
BenchmarkDecoder_DecodeAllParallel/kppkn.gtb.zst-16                       109733        108608        -1.03%
BenchmarkDecoder_DecodeAllParallel/geo.protodata.zst-16                   20827         20807         -0.10%
BenchmarkDecoder_DecodeAllParallel/plrabn12.txt.zst-16                    334782        333649        -0.34%
BenchmarkDecoder_DecodeAllParallel/lcet10.txt.zst-16                      251605        246310        -2.10%
BenchmarkDecoder_DecodeAllParallel/asyoulik.txt.zst-16                    85051         84388         -0.78%
BenchmarkDecoder_DecodeAllParallel/alice29.txt.zst-16                     121909        119929        -1.62%
BenchmarkDecoder_DecodeAllParallel/html_x_4.zst-16                        34743         34698         -0.13%
BenchmarkDecoder_DecodeAllParallel/paper-100k.pdf.zst-16                  4427          4395          -0.72%
BenchmarkDecoder_DecodeAllParallel/fireworks.jpeg.zst-16                  1253          1268          +1.20%
BenchmarkDecoder_DecodeAllParallel/urls.10K.zst-16                        218685        216112        -1.18%
BenchmarkDecoder_DecodeAllParallel/html.zst-16                            22815         22762         -0.23%
BenchmarkDecoder_DecodeAllParallel/comp-data.bin.zst-16                   1859          1869          +0.54%
```